Snowflake.Data/Client/SnowflakeDbCommand.cs

```diff
@@ -274,6 +281,179 @@ protected override async Task<DbDataReader> ExecuteDbDataReaderAsync(CommandBeha
+        public async Task<QueryStatus> GetQueryStatusAsync(string queryId, CancellationToken cancellationToken)
+        {
+            logger.Debug($"GetQueryStatusAsync");
+
+            // Check if queryId is valid uuid
+            if (UuidRegex.IsMatch(queryId))
+            {
+                var sfStatement = new SFStatement(connection.SfSession);
+                return await sfStatement.GetQueryStatusAsync(queryId, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                var errorMessage = $"The given query id {queryId} is not valid uuid";
+                logger.Error(errorMessage);
+                throw new Exception(errorMessage);
+            }
+        }
+
+        /// <summary>
+        /// Gets the query results based on query ID.
+        /// </summary>
+        /// <param name="queryId"></param>
+        /// <returns>The query results.</returns>
+        public DbDataReader GetResultsFromQueryId(string queryId)
+        {
+            logger.Debug($"GetResultsFromQueryId");
+
+            int retryPatternPos = 0;
+            int noDataCounter = 0;
+
+            QueryStatus status;
+            while (true)
+            {
+                status = GetQueryStatus(queryId);
+
+                if (!QueryStatuses.IsStillRunning(status))
+                {
+                    break;
+                }
+
+                // Timeout based on query status retry rules
+                Thread.Sleep(TimeSpan.FromSeconds(_asyncRetryPattern[retryPatternPos]));
+
+                // If no data, increment the no data counter
+                if (status == QueryStatus.NO_DATA)
+                {
+                    noDataCounter++;
+
+                    // Check if retry for no data is exceeded
+                    if (noDataCounter > AsyncNoDataMaxRetry)
+                    {
+                        var errorMessage = "Max retry for no data is reached";
+                        logger.Error(errorMessage);
+                        throw new Exception(errorMessage);
+                    }
+                }
+
+                if (retryPatternPos < _asyncRetryPattern.Length - 1)
+                {
+                    retryPatternPos++;
+                }
+            }
+
+            connection.SfSession.AsyncQueries.Remove(queryId);
+            SFBaseResultSet resultSet = sfStatement.GetResultWithId(queryId);
+
+            return new SnowflakeDbDataReader(this, resultSet);
+        }
+
+        /// <summary>
+        /// Gets the query results based on query ID.
+        /// </summary>
+        /// <param name="queryId"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>The query results.</returns>
+        public async Task<DbDataReader> GetResultsFromQueryIdAsync(string queryId, CancellationToken cancellationToken)
+        {
+            logger.Debug($"GetResultsFromQueryIdAsync");
+
+            int retryPatternPos = 0;

```

Comment:
It's extracted into a function. I'll check about moving it into a different class

Reply:
But let's try to refactor more. Would it be possible to remove isAsync parameter from the method signature and instead add to different parameters:
- `Func<string, QueryStatus> queryStatusVerifier`
- Action<int, CancellationToken> waitingAction

Then we could create 2 functions for sync and async scenario. One would provide functions:
- queryId => GetQueryStatus(queryId)
- (seonds, _) => Thread.Sleep(...);
And the other method would have the async versions.

Check if such a refactor is possible.

Reply:
I tried it and passing async functions has an error: `Cannot convert async lambda expression to delegate type 'Func<string, QueryStatus>'. An async lambda expression may return void, Task or Task<T>, none of which are convertible to 'Func<string, QueryStatus>'`
```
RetryUntilQueryResultIsAvailable(
    qId,
    cancellationToken,
    async queryId => await GetQueryStatusAsync(queryId, cancellationToken).ConfigureAwait(false),
    async (seconds, _) => await Task.Delay(TimeSpan.FromSeconds(seconds), cancellationToken).ConfigureAwait(false));
```

It's possible to turn the `Func` param to return a Task instead to make it work `Func<Task<QueryStatus>> queryStatusVerifier`

But afterwards I noticed some of the async tests finish much faster than they should because it wasn't actually waiting for the timeout. So I had to convert the `Action` param into `Func<int, CancellationToken, Task>` to call Wait() and that seems to fix it

I haven't committed these changes yet, I'll let you know when they're in

Reply:
I see that it gets a little complicated, so maybe let's not do this refactoring I suggested - we can leave it as it is now

Reply:
Ok